File: Structural - Facade/Python/facade/facade.py
from subsystem.subsystem_a import SubsystemA
from subsystem.subsystem_b import SubsystemB


class Facade:
    def __init__(self, subsystem_a: SubsystemA, subsystem_b: SubsystemB) -> None:
        self._subsystem_a = subsystem_a or SubsystemA()
        self._subsystem_b = subsystem_b or SubsystemB()

    def operation(self) -> str:
        results = []
        results.append("Facade initializes subsystems:")
        results.append(self._subsystem_a.operation1())
        results.append(self._subsystem_b.operation1())
        results.append("Facade orders subsystems to perform the action:")
        results.append(self._subsystem_a.operation_n())
        results.append(self._subsystem_b.operation_z())
        return "\n".join(results)

File: Structural - Facade/Typescript/facade/facade.ts
import { SubsystemA } from '../subsystem/subsystem-a.ts'
import { SubsystemB } from '../subsystem/subsystem-b.ts'


export class Facade {
  protected subsystemA: SubsystemA;

  protected subsystemB: SubsystemB;

  constructor(subsystemA?: SubsystemA, subsystemB?: SubsystemB) {
    this.subsystemA = subsystemA || new SubsystemA();
    this.subsystemB = subsystemB || new SubsystemB();
  }

  public operation(): string {
    let result = 'Facade initializes subsystems:\n';
    result += this.subsystemA.operation1();
    result += this.subsystemB.operation1();
    result += 'Facade orders subsystems to perform the action:\n';
    result += this.subsystemA.operationN();
    result += this.subsystemB.operationZ();

    return result;
  }
}